wip/beam-spec.md

```diff
@@ -0,0 +1,437 @@
+                }
+            }
+        }
+    }
+}
+```
+
+## Instructions
+
+The BEAM has a few special instructions that can be invoked using inline
+assembly `asm!`.
+
+- `asm!("ENABLE_INTERRUPTS")`. This instruction *enables* all interrupts.
+
+- `asm!("DISABLE_INTERRUPTS")`. This instruction *disables* all interrupts.
+
+- `asm!("SET_RUNNING_PRIORITY $0")`. This instruction changes the *running
+  priority* to the value `$0`.
+
+## Registers
+
+The BEAM has a special region of memory; write operations performed on this
+memory region can affect how interrupt handling works. Each significant piece of
+memory in this memory region is referred to as a *register*.
+
+- All registers are byte-sized
+
+- A program can only access the memory mapped registers by name using (part of)
+  the following `extern` block.
+
+``` rust
+extern "C" {
+    // write-only register
+    static MASK_INTERRUPT: UnsafeCell<u8>;
+
+    // write-only register
+    static UNMASK_INTERRUPT: UnsafeCell<u8>;
+
+    // read-write register
+    static PEND_INTERRUPT: UnsafeCell<u8>;
+
+    // read-only register
+    static RUNNING_PRIORITY: UnsafeCell<u8>;
+
+    // read-write register
+    static INTERRUPT_PRIORITY: [UnsafeCell<u8>; 8];
+}
+```
+
+- Performing *non*-volatile operations on these registers results in UNSPECIFIED
+  behavior.
+
+- Performing memory operations on these registers other than byte-sized reads /
+  writes results in UNSPECIFIED behavior.
+
+- The `MASK_INTERRUPT` register. Writing a value of `1` to bit `N` causes
+  interrupt `N` to be *masked*. Writing a value of `0` to any bit has no effect.
+  Reading this register returns an UNSPECIFIED value.
+
+- The `UNMASK_INTERRUPT` register. Writing a value of `1` to bit `N` causes
+  interrupt `N` to be *unmasked*. Writing a value of `0` to any bit has no
+  effect. Reading this register returns an UNSPECIFIED value.
+
+- The `PEND_INTERRUPT` register. Writing a value of `1` to bit `N` changes the
+  state of interrupt `N` to pending. Writing a value of `0` to any bit has no
+  effect. Reading this register returns the pending state of all interrupts; if
+  bit `N` has value `1` it means interrupt `N` is in the pending state; if bit
+  `N` has value `0` it means interrupt `N` is *not* in the pending state.
+
+- The `RUNNING_PRIORITY` register. Reading this register returns the current
+  *running priority*.
+
+- The `INTERRUPT_PRIORITY` registers. Each of these registers is used to control
+  the *priority* of the 8 interrupts the BEAM has. Writing a value `X` to the
+  `i`-th register changes the static priority of interrupt `i` to `X`. Reading
+  the `i`-th register returns the static priority of interrupt `i`.
+
+## Examples
+
+This section contains examples (real Rust code) that target the BEAM. They are
+meant to improve your understanding of the above specification but they are not
+part of the specification itself.
+
+In all these examples the BEAM goes from state A, to B, to C, etc.
```

Comment:
This sentence makes no sense unless one has already read an example. Maybe announce in advance there will be comments detailing the states the machine can be in?